haystack/dataclasses/chat_message.py

```diff
@@ -27,11 +47,93 @@ class ChatMessage:
+        elif isinstance(self.content, list):
+            content: List[Union[str, ByteStream]] = []
+            content_types = []
+            for part in self.content:
+                if isinstance(part, str):
+                    part_content, content_type = self._parse_string_content(part)
+                    content.append(part_content)
+                    content_types.append(content_type)
+                elif isinstance(part, ByteStream):
+                    part_content, content_type = self._parse_byte_stream_content(part)
+                    content.append(part_content)
+                    content_types.append(content_type)
+                else:
+                    raise ValueError("Invalid element in content. Valid types are str and ByteStream" "Element: {part}")
+            self.content = content
+            self.meta["__haystack_content_type__"] = content_types
+        else:
+            raise ValueError(
+                f"Invalid type of content. Valid types are str, "
+                f"ByteStream and a list of str and ByteStream objects."
+                f"Content: {self.content}"
+            )
+
+    @staticmethod
+    def _parse_string_content(content: str) -> Tuple[str, ContentType]:
+        """
+        Parse the string content to differentiate between different types of string representable content.
+
+        :param content: The string content you want to parse.
+        :returns: a tuple containing the parsed content and the content type.
+        """
+        if content.strip().startswith("image_url:"):
+            url = content.split("image_url:")[-1].strip()
+            return url, ContentType.IMAGE_URL
+        else:  # TEXT
+            return content, ContentType.TEXT
+
+    @staticmethod
+    def _parse_byte_stream_content(content: ByteStream) -> Tuple[ByteStream, ContentType]:
+        """
+        Parse the byte stream content to differentiate between different types of byte encoded content.
+
+        :param content: The bytes content you want to parse.
+        :returns: a tuple containing the parsed content and the content type.
+
+        :raises: Value error if the 'mime_type' attribute is None or any invalid value.
+        """
+        if content.mime_type is None:
+            raise ValueError(
+                "Unidentified ByteStream added as part of the content of the ChatMessage."
+                "Populate thee 'mime_type' attribute with the identifier of the content type."
+            )
+
+        content_type = ContentType(content.mime_type)
+        if not content_type.is_valid_byte_stream_type():
+            raise ValueError(
+                f"The 'mime_type' attribute of the introduced content "
+                f"has a not valid ContentType for a ByteStream"
+                f"Value: {content_type}. Valid content types:"
+                + ", ".join([c.value for c in ContentType.valid_byte_stream_types()])
+            )
+
+        return content, content_type
```

Comment:
I would replace this code with a resolver that takes in the ByteStream object and returns a ContentType, else raises a ValueError in case the ByteStream object is invalid, e.g.
```
if mime_type.startswith("image/"):
    return ContentType("IMAGE_BASE64")
if mime_type.startswith("audio/"):
    return ContentType("AUDIO")
raise ValueError("...")
```

Then the `to_openai_format` can have a single process for any type of image, dynamically filling the message dictionary with the `mime_type` of the `ByteStream` object. This avoids a separate process for every different kind of content, which is going to multiply in the future.